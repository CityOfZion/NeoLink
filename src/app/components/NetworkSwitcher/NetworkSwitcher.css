.networkNavigation {
  position: relative;
  height: 100%;
  width: 25%;
}

.networkNavigationButtonContainer {
  height: 100%;
}

.networkNavigationButton {
  width: 100%;
  display: flex;
  background-color: transparent;
  border: none;
  justify-content: center;
  height: 100%;
  cursor: pointer;
}

.networkNavigationButtonContent {
  border: 1px solid #fff;
  border-radius: 25px;
  padding: 6px 12px;
  margin-right: 15px;
  color: #fff;
  display: flex;
  justify-content: center;
  align-items: center;
  pointer-events: none;
  min-width: 85px;
  max-width: 85px;
}

.networkNavigationChevron {
  width: 10px;
  transform: translateY(2px);
  margin-left: 5px;
}

.networkNavigationDropDownContainer {
  padding: 15px 5px;
  width: 140px;
  left: -77px;
  top: 49px;
}

.networkOptionButton {
  display: block;
  width: 100%;
  margin: 6px 0;
  border: none;
  font-family: Roboto;
  background: transparent;
  color: #585858;
  cursor: pointer;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  font-size: 0.8rem;
}

.networkOptionButton:first-child {
  margin-top: 0;
}

.networkOptionButton:last-child {
  margin-bottom: 0;
}

.networkNavigationOptionSelected {
  background-color: #48f12f;
  border: none;
  width: 7px;
  height: 7px;
  border-radius: 50%;
  margin-left: auto;
}

.mainNetNeoImg,
.networkOptionIcon {
  width: 16px;
  margin-right: 8px;
}

@keyframes fadeInMenu {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}
