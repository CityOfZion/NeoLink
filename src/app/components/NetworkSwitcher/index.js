import React, { Component } from 'react'
import PropTypes from 'prop-types'

import style from './NetworkSwitcher.css'

import globe from '../../../img/globe.svg'
import chevron from '../../../img/chevron-down.svg'
import neoImg from '../../../img/icon-34.png'
import flask from '../../../img/flask.svg'

class NetworkSwitcher extends Component {
  state = {
    networkMenuOpen: false,
  }

  componentDidMount() {
    window.addEventListener('click', this.closeDropdownMenu)
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.closeDropdownMenu)
  }

  changeNetwork = event => {
    const { setNetwork } = this.props
    let dataset = event.target.dataset.value || event.target.parentNode.dataset.value

    if (dataset) {
      setNetwork(dataset)
      this.closeDropdownMenu()
    }
  }

  toggleDropdownMenu = () => {
    this.setState({ networkMenuOpen: !this.state.networkMenuOpen })
  }

  closeDropdownMenu = event => {
    if (!event.target.className.includes('NetworkSwitcher')) {
      this.setState({ networkMenuOpen: false })
    }
  }

  getIndicator = (networks, index) => {
    let indicator
    const networkName = networks[index].name

    if (networkName === 'MainNet') {
      indicator = <img src={ neoImg } alt='neo' className={ style.mainNetNeoImg } />
    } else if (networkName === 'TestNet') {
      indicator = <img src={ flask } alt='flask' className={ style.networkOptionIcon } />
    } else {
      indicator = (
        <div
          style={ {
            height: '13px',
            width: '15px',
            marginRight: '8px',
            borderRadius: '3px',
            backgroundColor: '#f15c5c',
          } }
        />
      )
    }

    return indicator
  }

  generateNetworkOptions() {
    const networkOptions = []
    const { selectedNetworkId, networks } = this.props

    Object.keys(networks).forEach(index => {
      const indicator = this.getIndicator(networks, index)

      const selected = selectedNetworkId === networks[index].name

      networkOptions.push(
        <button key={ `option-key-${index}` } data-value={ index } className={ style.networkOptionButton }>
          {indicator}
          {networks[index].name}
          {selected && <div className={ style.networkNavigationOptionSelected } />}
        </button>
      )
    })
    return networkOptions
  }

  generateDropdownClasses = () => {
    const { networkMenuOpen } = this.state

    let dropdownStyles
    if (networkMenuOpen) {
      dropdownStyles = `${style.networkNavigationDropdown} ${style.networkNavigationDropdownOpen}`
    } else {
      dropdownStyles = style.networkNavigationDropdown
    }

    return dropdownStyles
  }

  render() {
    const networkOptions = this.generateNetworkOptions()
    const dropdownStyles = this.generateDropdownClasses()

    return (
      <section className={ style.networkNavigation }>
        <button className={ style.networkNavigationButton } onClick={ this.toggleDropdownMenu }>
          <img src={ globe } className={ style.networkNavigationGlobe } alt='globe' />
          <img src={ chevron } className={ style.networkNavigationChevron } alt='chevron down' />
        </button>
        <div className={ dropdownStyles } onClick={ this.changeNetwork }>
          {networkOptions}
        </div>
      </section>
    )
  }
}

NetworkSwitcher.propTypes = {
  selectedNetworkId: PropTypes.string,
  setNetwork: PropTypes.func,
  networks: PropTypes.object,
}

export default NetworkSwitcher
