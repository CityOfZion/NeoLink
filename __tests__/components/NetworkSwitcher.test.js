import React from 'react'

import { mount } from 'enzyme'

import NetworkSwitcher from '../../src/app/components/NetworkSwitcher'

const setup = () => {
  const props = {
    selectedNetworkId: 'MainNet',
    account: {
      address: 'ARjkxk6VcKPFKqRHhuLNog9TbdYxhKu9be',
    },
    networks: {
      MainNet: { name: 'MainNet', url: 'http://api.wallet.cityofzion.io', canDelete: false, apiType: 'neoscan' },
      TestNet: {
        name: 'TestNet',
        url: 'http://testnet-api.wallet.cityofzion.io',
        canDelete: false,
        apiType: 'neoscan',
      },
      CoZTestNet: {
        name: 'CoZ TestNet',
        url: 'http://coz-privatenet.herokuapp.com/',
        canDelete: false,
        apiType: 'neoscan',
      },
    },
    setNetwork: jest.fn(),
    getBalance: jest.fn(),
  }

  const wrapper = mount(
    <NetworkSwitcher { ...props } />
  )

  return {
    wrapper,
  }
}

describe('NetworkSwitch', () => {
  // test('Renders without crashing', () => {
  //   const { wrapper } = setup()
  //   expect(wrapper).toMatchSnapshot()
  // })
  test('correctly renders MainNet initially', () => {
    const { wrapper } = setup()
    wrapper.find('.dropDownButton').simulate('click')
    const networkSelectorElement = wrapper.find('.networkOptionButton').at(0)

    expect(networkSelectorElement.contains(<div className='networkNavigationOptionSelected' />)).toBe(true)
  })

  test('switches to the correct network when chosen from the dropdown', async () => {
    const { wrapper } = setup()

    const instance = wrapper.instance()
    wrapper.find('.dropDownButton').simulate('click')
    let networkSelectorElement = wrapper.find('.networkOptionButton').at(1)
    networkSelectorElement.simulate('click')

    expect(instance.props.setNetwork).toHaveBeenCalledWith('TestNet')

    networkSelectorElement = wrapper.find('.networkOptionButton').at(0)
    networkSelectorElement.simulate('click')

    expect(instance.props.setNetwork).toHaveBeenCalledWith('MainNet')
  })
})
