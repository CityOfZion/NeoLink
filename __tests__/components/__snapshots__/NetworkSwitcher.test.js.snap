// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkSwitch renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NetworkSwitcher
    networks={
      Object {
        "CoZTestNet": Object {
          "canDelete": false,
          "name": "CoZ TestNet",
          "url": "http://coz-privatenet.herokuapp.com/",
        },
        "MainNet": Object {
          "canDelete": false,
          "name": "MainNet",
          "url": "http://api.wallet.cityofzion.io",
        },
        "TestNet": Object {
          "canDelete": false,
          "name": "TestNet",
          "url": "http://testnet-api.wallet.cityofzion.io",
        },
      }
    }
    selectedNetworkId="MainNet"
    setNetwork={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <select
        className="switcher"
        defaultValue="MainNet"
        onChange={[Function]}
      >
        <option
          value="MainNet"
        >
          MainNet
        </option>
        <option
          value="TestNet"
        >
          TestNet
        </option>
        <option
          value="CoZTestNet"
        >
          CoZ TestNet
        </option>
      </select>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <option
            value="MainNet"
          >
            MainNet
          </option>,
          <option
            value="TestNet"
          >
            TestNet
          </option>,
          <option
            value="CoZTestNet"
          >
            CoZ TestNet
          </option>,
        ],
        "className": "switcher",
        "defaultValue": "MainNet",
        "onChange": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "option-key-MainNet",
          "nodeType": "host",
          "props": Object {
            "children": "MainNet",
            "value": "MainNet",
          },
          "ref": null,
          "rendered": "MainNet",
          "type": "option",
        },
        Object {
          "instance": null,
          "key": "option-key-TestNet",
          "nodeType": "host",
          "props": Object {
            "children": "TestNet",
            "value": "TestNet",
          },
          "ref": null,
          "rendered": "TestNet",
          "type": "option",
        },
        Object {
          "instance": null,
          "key": "option-key-CoZTestNet",
          "nodeType": "host",
          "props": Object {
            "children": "CoZ TestNet",
            "value": "CoZTestNet",
          },
          "ref": null,
          "rendered": "CoZ TestNet",
          "type": "option",
        },
      ],
      "type": "select",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <select
          className="switcher"
          defaultValue="MainNet"
          onChange={[Function]}
        >
          <option
            value="MainNet"
          >
            MainNet
          </option>
          <option
            value="TestNet"
          >
            TestNet
          </option>
          <option
            value="CoZTestNet"
          >
            CoZ TestNet
          </option>
        </select>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              value="MainNet"
            >
              MainNet
            </option>,
            <option
              value="TestNet"
            >
              TestNet
            </option>,
            <option
              value="CoZTestNet"
            >
              CoZ TestNet
            </option>,
          ],
          "className": "switcher",
          "defaultValue": "MainNet",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "option-key-MainNet",
            "nodeType": "host",
            "props": Object {
              "children": "MainNet",
              "value": "MainNet",
            },
            "ref": null,
            "rendered": "MainNet",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "option-key-TestNet",
            "nodeType": "host",
            "props": Object {
              "children": "TestNet",
              "value": "TestNet",
            },
            "ref": null,
            "rendered": "TestNet",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "option-key-CoZTestNet",
            "nodeType": "host",
            "props": Object {
              "children": "CoZ TestNet",
              "value": "CoZTestNet",
            },
            "ref": null,
            "rendered": "CoZ TestNet",
            "type": "option",
          },
        ],
        "type": "select",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
